### ISSUE TITLE
Problems caused by /etc/localtime symlink

### ISSUE THREAD

**Issue by mat8913** [flatpak/flatpak#3338]
I've been trying to find the root cause of the incorrect handling of /etc/localtime causing problems with Steam chat, FreeTube, Riot and more. I think I've traced the problem back to libicu, but before I report it there I want to get your opinion on this possible hangup.

[`man 5 localtime`](http://man7.org/linux/man-pages/man5/localtime.5.html) has the following paragraph:

> The /etc/localtime file configures the system-wide timezone of the local system that is used by applications for presentation to the user. It should be an absolute or relative symbolic link pointing to /etc/zoneinfo/, followed by a timezone identifier such as "Europe/Berlin" or "Etc/UTC".

This seems to imply that flatpak is symlinking localtime incorrectly since:

```
[📦 org.freedesktop.Platform ~]$ ls -l /etc/localtime 
lrwxrwxrwx 1 matthew matthew 27 Dec 24 14:38 /etc/localtime -> /run/host/monitor/localtime
[📦 org.freedesktop.Platform ~]$ ls -l /run/host/monitor/localtime
-rw-r--r-- 1 matthew matthew 2204 Dec 22 20:40 /run/host/monitor/localtime
```

Is that section of the manual considered authoritative, or is it just something the systemd developers made up?

---

In any case, here's a simple C++ program to demonstrate one manifestation of the problem:

```
#include <iostream>
#include <unicode/timezone.h>


int main() {
	icu::TimeZone* hostZone = icu::TimeZone::detectHostTimeZone();

	icu::UnicodeString r1;
	hostZone->getDisplayName(r1);

	std::string r2;
	r1.toUTF8String(r2);

	std::cout << r2 << std::endl;
	return 0;
}
```

Compile with `g++ -licui18n -licuuc main.cpp`. Outside the flatpak sandbox, it gives the correct result `Australian Eastern Standard Time`, whereas inside the sandbox it gives `GMT+10:00` which is incorrect because we're currently in daylight saving.

**Comment by refi64** [comment]
I decided to look into this a tad, and it seems on-point. libicu [resolves the localtime symlink](https://github.com/unicode-org/icu/blob/d1293c50cbef31156f3d28f16cc67264389fd8fc/icu4c/source/common/putil.cpp#L1113) and then [searches `/usr/lib/zoneinfo` for a file with the same contents](https://github.com/unicode-org/icu/blob/d1293c50cbef31156f3d28f16cc67264389fd8fc/icu4c/source/common/putil.cpp#L1134). (Side note, most distros have /etc/zoneinfo as a symlink to /usr/share/zoneinfo.)

**Comment by alexlarsson** [comment]
This is tricky, because we really want to use the host zoneinfo file which is more likely to be up-to-date than the one in the runtime, and also because it may change on the host.

We could resolve the host symlink at startup and make /etc/localtime be a symlink to /etc/zoneinfo/Europe/Berlin which is a symlink to /run/host/monitor/localtime.

Then if you *opened* the /etc/localtime file you would get whatever is the current timezone data on the host, but if you look at the symlink name you would only ever get whatever was current when the app started.

**Comment by refi64** [comment]
Could /usr/share/zoneinfo be a symlink into /run/host or similar? Using /etc wouldn't fix the libicu case, since it looks into /usr/share exclusively (unless /usr/share/zoneinfo is a symlink to /etc/zoneinfo).

**Comment by alexlarsson** [comment]
I'm not sure how standardized /usr/share/zoneinfo is. Can we rely on all distros to have them in that place in a format that is ABI stable forever?

**Comment by damianatorrpm** [comment]
Could this be related: https://gitlab.gnome.org/GNOME/gnome-calendar/issues/485

**Comment by nanonyme** [comment]
Note Steam app has a workaround where it searches /usr/share/zoneinfo itself for /etc/localtime, then sets TZ to disable the search logic in various libraries. This has more or less worked except you might get equivalent time zone instead of real time zone.

**Comment by refi64** [comment]
Okay so I was just looking into this again, time to make some corrections since I previously looked into this late at night while also trying to read the libicu source code on a phone:

First off, there is no `/etc/zoneinfo`, I was confusing it with something else (yeah this one's just on me).

Here's the more interesting part: libicu actually has *two* different ways of finding the timezone. If `/etc/localtime` is a symlink, this actually takes a [*completely different code path*](https://github.com/unicode-org/icu/blob/d1293c50cbef31156f3d28f16cc67264389fd8fc/icu4c/source/common/putil.cpp#L1113-L1123) that reads the link target of `/etc/localtime` and then uses whatever is after `/zoneinfo/` as the timezone name.

This does mean that some sort of indirection like @alexlarsson had originally mentioned *should* work...but in that case, if the timezone is updated, the application will find the new timezone data under the *original* timezone's name, which could be incredibly confusing on the user's end. I really can't think of any way to reliably work around this using the current setup, short of using some ptrace stuff or similar like fakeroot/fakeroot-ng/proot do, binary patching libicu, ld_preload, or other nasty hacks that probably wouldn't work out incredibly well here.

Could there be some way of giving the system helper /etc's file descriptor, and then the system helper could adjust the localtime symlink by itself using openat/symlinkat/renameat? I did a [quick, ugly Python script as a demo](https://gist.github.com/refi64/953e06d85db2a13ce7e99fd3061b294c) of manipulating symlinks under a directory fd sent by a child process in a sandbox, and it seems to work... At that point, Flatpak could mount the host's /usr/share/zoneinfo onto the same location of the sandbox.

In addition, this *also* means that, on a system where /etc/localtime is *not* a symlink and the system monitor is *not* present, libicu would again fail to find the time zone if the runtime's `/usr/share/zoneinfo` differs from the hosts. I believe this could be easily worked around by always bind mounting the host's localtime onto the sandbox's `/usr/share/zoneinfo` subpath. (Worth noting that I *think* the current code has a small bug: if the system monitor is not present, /etc/localtime is a symlink, and the zoneinfo file is on the host but not in the sandbox, the sandbox's /etc/localtime will be a broken symlink. A fix for this would be to do the same thing, bind mounting the host's localtime file onto the sandbox's `/usr/share/zoneinfo` subpath. So, these two cases would be able to take the same code paths.)

TL;DR: [relevant xkcd](https://xkcd.com/1883/).

**Comment by BailShamber** [comment]
Related bugs:
https://bugzilla.mozilla.org/show_bug.cgi?id=1642496
https://github.com/flathub/com.skype.Client/issues/88

**Comment by alexlarsson** [comment]
So, it seems that the tzdata files sometimes do change format:
https://developers.redhat.com/blog/2019/02/22/time-zone-data-tzdata-2018-data-format-changes-and-red-hat-enterprise-linux/
This makes it somewhat risky to expose to containerized apps.

**Comment by fweimer** [comment]
@alexlarsson This was technically not a format change as such, it is about what applications do with the stored data.

This issue here looks very much like [this bug](https://bugzilla.redhat.com/show_bug.cgi?id=1399749). Flatpaks may behave differently from production-grade container images, so [my recommendation](https://bugzilla.redhat.com/show_bug.cgi?id=1399749#c30) may not apply to them.

If you want a quick fix, you will have to rewrite the target of the `/etc/localtime` symbolic link *and* bind-mount `/usr/share/zoneinfo`. The paths should be consistent across Linux distributions. The manpage quoted initially is *very* outdated because it mentions `/etc/zoneinfo`. If you only update one of those, you will encounter issues related to differences in supported time zone names.

Older Red Hat Enterprise Linux 6 systems are an outlier because [symbolic links are automatically overwritten based on data from `/etc/sysconfig/clock`](https://bugzilla.redhat.com/show_bug.cgi?id=1200555), but that was fixed in Red Hat Enterprise Linux 6.8. To support these old systems, it may also be necessary to update `/etc/sysconfig/clock` along with `/etc/localtime`.
